drivers/w1/w1_zephyr_gpio.c

```diff
@@ -0,0 +1,326 @@
+ */
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/w1.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(w1_gpio, CONFIG_W1_LOG_LEVEL);
+
+/*
+ * The time critical sections are used to ensure that the timing
+ * between communication operations is correct.
+ */
+#if defined(CONFIG_W1_ZEPHYR_GPIO_TIME_CRITICAL)
+#define W1_GPIO_ENTER_CRITICAL()   irq_lock()
+#define W1_GPIO_EXIT_CRITICAL(key) irq_unlock(key)
+#define W1_GPIO_WAIT_US(us)        k_busy_wait(us)
+#else
+#define W1_GPIO_ENTER_CRITICAL()   0u
+#define W1_GPIO_EXIT_CRITICAL(key) (void)key
+#define W1_GPIO_WAIT_US(us)        k_usleep(us)
+#endif
+
+/*
+ * Standard timing between communication operations:
+ */
+#define W1_GPIO_TIMING_STD_A 6u
+#define W1_GPIO_TIMING_STD_B 64u
+#define W1_GPIO_TIMING_STD_C 60u
+#define W1_GPIO_TIMING_STD_D 10u
+#define W1_GPIO_TIMING_STD_E 9u
+#define W1_GPIO_TIMING_STD_F 55u
+#define W1_GPIO_TIMING_STD_G 0u
+#define W1_GPIO_TIMING_STD_H 480u
+#define W1_GPIO_TIMING_STD_I 70u
+#define W1_GPIO_TIMING_STD_J 410u
+
+/*
+ * Overdrive timing between communication operations:
+ *
+ * Not completely correct since the overdrive communication requires
+ * delays of 2.5us, 7.5us and 8.5us.
+ * The delays are approximated by flooring the values.
+ */
+#define W1_GPIO_TIMING_OD_A 1u
+#define W1_GPIO_TIMING_OD_B 7u
+#define W1_GPIO_TIMING_OD_C 7u
+#define W1_GPIO_TIMING_OD_D 2u
+#define W1_GPIO_TIMING_OD_E 1u
+#define W1_GPIO_TIMING_OD_F 7u
+#define W1_GPIO_TIMING_OD_G 2u
+#define W1_GPIO_TIMING_OD_H 70u
+#define W1_GPIO_TIMING_OD_I 8u
+#define W1_GPIO_TIMING_OD_J 40u
+
+struct w1_gpio_timing {
+	uint16_t a;
+	uint16_t b;
+	uint16_t c;
+	uint16_t d;
+	uint16_t e;
+	uint16_t f;
+	uint16_t g;
+	uint16_t h;
+	uint16_t i;
+	uint16_t j;
+};
+
+struct w1_gpio_config {
+	/** w1 master config, common to all drivers */
+	struct w1_master_config master_config;
+	/** GPIO device used for 1-Wire communication */
+	const struct gpio_dt_spec spec;
+};
+
+struct w1_gpio_data {
+	/** w1 master data, common to all drivers */
+	struct w1_master_data master_data;
+	/** timing parameters for 1-Wire communication */
+	const struct w1_gpio_timing *timing;
+	/** overdrive speed mode active */
+	bool overdrive_active;
+};
+
+static struct w1_gpio_timing std = {
+	.a = W1_GPIO_TIMING_STD_A,
+	.b = W1_GPIO_TIMING_STD_B,
+	.c = W1_GPIO_TIMING_STD_C,
+	.d = W1_GPIO_TIMING_STD_D,
+	.e = W1_GPIO_TIMING_STD_E,
+	.f = W1_GPIO_TIMING_STD_F,
+	.g = W1_GPIO_TIMING_STD_G,
+	.h = W1_GPIO_TIMING_STD_H,
+	.i = W1_GPIO_TIMING_STD_I,
+	.j = W1_GPIO_TIMING_STD_J,
+};
+
+static struct w1_gpio_timing od = {
+	.a = W1_GPIO_TIMING_OD_A,
+	.b = W1_GPIO_TIMING_OD_B,
+	.c = W1_GPIO_TIMING_OD_C,
+	.d = W1_GPIO_TIMING_OD_D,
+	.e = W1_GPIO_TIMING_OD_E,
+	.f = W1_GPIO_TIMING_OD_F,
+	.g = W1_GPIO_TIMING_OD_G,
+	.h = W1_GPIO_TIMING_OD_H,
+	.i = W1_GPIO_TIMING_OD_I,
+	.j = W1_GPIO_TIMING_OD_J,
+};
```

Comment:
This may save some RAM:

```suggestion
static const struct w1_gpio_timing std = {
	.a = W1_GPIO_TIMING_STD_A,
	.b = W1_GPIO_TIMING_STD_B,
	.c = W1_GPIO_TIMING_STD_C,
	.d = W1_GPIO_TIMING_STD_D,
	.e = W1_GPIO_TIMING_STD_E,
	.f = W1_GPIO_TIMING_STD_F,
	.g = W1_GPIO_TIMING_STD_G,
	.h = W1_GPIO_TIMING_STD_H,
	.i = W1_GPIO_TIMING_STD_I,
	.j = W1_GPIO_TIMING_STD_J,
};

static const struct w1_gpio_timing od = {
	.a = W1_GPIO_TIMING_OD_A,
	.b = W1_GPIO_TIMING_OD_B,
	.c = W1_GPIO_TIMING_OD_C,
	.d = W1_GPIO_TIMING_OD_D,
	.e = W1_GPIO_TIMING_OD_E,
	.f = W1_GPIO_TIMING_OD_F,
	.g = W1_GPIO_TIMING_OD_G,
	.h = W1_GPIO_TIMING_OD_H,
	.i = W1_GPIO_TIMING_OD_I,
	.j = W1_GPIO_TIMING_OD_J,
};
```